### PR TITLE
Search ancestor and its references for default projects

### LINKED ISSUE #56959: Project references: TS server does not find correct project
### 🔎 Search Terms

- project references
- tsconfig.json

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about project references.

### ⏯ Playground Link

_No response_

### 💻 Code

Full reduced test case: https://github.com/OliverJAsh/ts-project-references-nested

Contents inlined below.

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "outDir": "./dist/"
  },
  "references": [
    { "path": "./demos/tsconfig.json" },
    { "path": "./app/tsconfig.json" }
  ]
}

```
`app/tsconfig.json`:
```json
{
  "compilerOptions": {
    "composite": true,
    "outDir": "../app-dist/"
  },
  "include": ["**/*"],
  "exclude": ["**/*-demos.*"]
}

```
`app/Component-demos.ts`:
```ts
// This errors.
import * as helpers from 'demos/helpers';

export const demo = () => {
    helpers;
}

```
`demos/tsconfig.json`:
```json
{
  "compilerOptions": {
    "composite": true,
    "rootDir": "../",
    "outDir": "../demos-dist/",
    "paths": {
      "demos/*": ["./*"]
    }
  },
  "include": [
    "**/*",
    "../app/**/*-demos.*"
  ],
  "references": [{ "path": "../app/tsconfig.json" }]
}

```
`demos/helpers.ts`:
```ts
export const foo = 1;

```

### 🙁 Actual behavior

When I run `tsc --build --verbose tsconfig.json`, there are no errors, but when I open my editor (VS Code) and open the file `app/Component-demos.ts`, I see errors:

image

### 🙂 Expected behavior

No errors inside my editor.

### Additional information about the issue

When `app/Component-demos.ts` is opened, I believe TypeScript traverses parent file directories searching for a `tsconfig.json` file, and it stops at the first file it finds which in this case is `app/tsconfig.json`. This TS project does not include the file `app/Component-demos.ts`, so I believe TypeScript should continue traversing parent file directories to search for another `tsconfig.json` that does include the file. In this case, we want TypeScript to discover the root `tsconfig.json`.

### PR DESCRIPTION
**Pull request by sheetalkamat**
On Top of #58486

When finding default project for open file, we search tsconfig .. Before this change, we would stop at first tsconfig/jsconfig encountered and check if any of references has the file. If none of this had file we would stop and put file in inferred projects. 
This has issues when you arrange code such that your test files are next to source files and are not using project references or the default project is the project for building sources. So test files endup in inferred project and dont get their custom options. 

With this change, 

- Continue search for the tsconfig further if we couldnt find default project as if we didnt find this tsconfig (so till projectRootPath) and try to find the default project.
- Caches the config file name for ancestor lookups as well as we will be skipping to config if it doesn't contain project
- Fixed config search of ancestor lookup to not skip jsconfig if tsconfig is present

Fixes #56959

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot run dt
@typescript-bot perf test this
@typescript-bot pack this

**Comment by sheetalkamat** [comment]
Ping @andrewbranch @jakebailey to review so we can get this in nightly and see how it goes

File: src/server/editorServices.ts [hunk 1]
```diff
@@ -708,7 +739,8 @@ function forEachAncestorProject<T>(
         if (
             !project.isInitialLoadPending() &&
             (
-                !project.getCompilerOptions().composite ||
+                (searchOnlyPotentialSolution && !project.getCompilerOptions().composite) ||
+                // TODO: Should this flag be shared between trying to load solution for find all references vs trying to find default project
```
Comment (jakebailey) [inline src/server/editorServices.ts:743]: Is this still something you're looking to address?

File: tests/baselines/reference/tsserver/projects/file-opened-is-in-configured-project-that-will-be-removed.js [hunk 2]
```diff
@@ -406,42 +406,18 @@ Info seq  [hh:mm:ss:mss] event:
         "diagnostics": []
       }
     }
-Info seq  [hh:mm:ss:mss] `remove Project::
+Info seq  [hh:mm:ss:mss] getConfigFileNameForFile:: File: /user/username/projects/myproject/playground/tsconfig-json/tsconfig.json ProjectRootPath: undefined:: Result: /user/username/projects/myproject/playground/tsconfig.json
 Info seq  [hh:mm:ss:mss] Project '/user/username/projects/myproject/playground/tsconfig.json' (Configured)
 Info seq  [hh:mm:ss:mss] 	Files (4)
```
Comment (jakebailey) [inline tests/baselines/reference/tsserver/projects/file-opened-is-in-configured-project-that-will-be-removed.js:411]: I assume the changes here are positive, in that it's finding another reasonable project for the file?

Comment (jakebailey) [reply tests/baselines/reference/tsserver/projects/file-opened-is-in-configured-project-that-will-be-removed.js:411]: (Not 100% sure what "will be removed" means for these tests; this folder has a couple of tests that change like this.)

Comment (sheetalkamat) [reply tests/baselines/reference/tsserver/projects/file-opened-is-in-configured-project-that-will-be-removed.js:411]: Earlier we were removing the project from memory that means free up sources, but now the file is part of the project 
Info seq  [hh:mm:ss:mss] 	FileName: /user/username/projects/myproject/playground/tsconfig-json/tests/spec.ts ProjectRootPath: undefined
Info seq  [hh:mm:ss:mss] 		Projects: /user/username/projects/myproject/playground/tsconfig.json

**Review by jakebailey — APPROVED** [review]
LGTM besides the TODO; not sure what you want to do with that.

File: src/server/editorServices.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/server/editorServices.ts:743]: I really dont know if different option is needed or not. For now i will go with this as is and we can add option later if we see people asking for it?

Comment (sheetalkamat) [reply src/server/editorServices.ts:743]: Will modify the comment to reflect that

**Comment by qMatr1xp** [comment]
projRef

**Comment by qMatr1xp** [comment]
[](url)

**Pull request merged**